vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -0,0 +1,453 @@
+                    offset_[iseg].second);
+            for (size_t iy = 0; iy < cost_map_[iseg].dim_size(1); iy++) {
+                for (size_t ix = 0; ix < cost_map_[iseg].dim_size(0); ix++) {
+                    fprintf(fp, "%.4g,\t",
+                            cost_map_[iseg][ix][iy].delay);
+                }
+                fprintf(fp, "\n");
+            }
+            fprintf(fp, "\n\n");
+        }
+
+        fclose(fp);
+    }
+
+  private:
+    std::vector<vtr::NdMatrix<Cost_Entry, 2>> cost_map_;
+    std::vector<std::pair<int, int>> offset_;
+    std::vector<int> segment_map_;
+};
+
+static CostMap g_cost_map;
+
+class StartNode {
+  public:
+    StartNode(int start_x, int start_y, t_rr_type rr_type, int seg_index)
+        : start_x_(start_x)
+        , start_y_(start_y)
+        , rr_type_(rr_type)
+        , seg_index_(seg_index)
+        , index_(0) {}
+    int get_next_node() {
+        const auto& device_ctx = g_vpr_ctx.device();
+        const std::vector<int>& channel_node_list = device_ctx.rr_node_indices[rr_type_][start_x_][start_y_][0];
+
+        for (; index_ < channel_node_list.size(); index_++) {
+            int node_ind = channel_node_list[index_];
+
+            if (device_ctx.rr_nodes[node_ind].capacity() == 0) {
+                continue;
+            }
+
+            const std::pair<size_t, size_t>* loc = device_ctx.connection_boxes.find_canonical_loc(node_ind);
+            if (loc == nullptr) {
+                continue;
+            }
+
+            int node_cost_ind = device_ctx.rr_nodes[node_ind].cost_index();
+            int node_seg_ind = device_ctx.rr_indexed_data[node_cost_ind].seg_index;
+            if (node_seg_ind == seg_index_) {
+                index_ += 1;
+                return node_ind;
+            }
+        }
+
+        return UNDEFINED;
+    }
+
+  private:
+    int start_x_;
+    int start_y_;
+    t_rr_type rr_type_;
+    int seg_index_;
+    size_t index_;
+};
+
+void compute_connection_box_lookahead(
+    const std::vector<t_segment_inf>& segment_inf) {
+    size_t num_segments = segment_inf.size();
+    vtr::ScopedStartFinishTimer timer("Computing connection box lookahead map");
+
+    /* free previous delay map and allocate new one */
+    g_cost_map.set_segment_count(segment_inf.size());
+
+    /* run Dijkstra's algorithm for each segment type & channel type combination */
+    for (int iseg = 0; iseg < (ssize_t)num_segments; iseg++) {
+        VTR_LOG("Creating cost map for %s(%d)\n",
+                segment_inf[iseg].name.c_str(), iseg);
+        /* allocate the cost map for this iseg/chan_type */
+        t_routing_cost_map cost_map;
+
+        int count = 0;
+
+        int dx = 0;
+        int dy = 0;
+        //int start_x = vtr::nint(device_ctx.grid.width()/2);
+        //int start_y = vtr::nint(device_ctx.grid.height()/2);
+        int start_x = REF_X;
+        int start_y = REF_Y;
+        while ((count == 0 && dx < 7) || dy < 1) {
```

Comment:
How `7` and `1` are chosen? I suppose it would be better to use some defines.

Reply:
Done